File: binary-search-medium/153.寻找旋转排序数组中的最小值.c
/*
 * @lc app=leetcode.cn id=153 lang=c
 *
 * [153] 寻找旋转排序数组中的最小值
 */
#include <limits.h>
// @lc code=start
int Min(int a, int b) { return a < b ? a : b; }
// int findMin(int* nums, int numsSize) {
//   int min = INT_MAX, left = 0, mid, right = numsSize - 1;
//   while (left <= right) {
//     mid = left + (right - left) / 2;
//     min = Min(min, nums[mid]);
//     if (nums[left] <= nums[mid] && nums[mid] < nums[right])
//       right = mid - 1;
//     else if (nums[left] <= nums[mid] && nums[mid] > nums[right]) {
//       if (nums[left] > nums[right])
//         left = mid + 1;
//       else
//         right = mid - 1;
//     } else if (nums[left] > nums[mid] && nums[right] > nums[mid])
//       right = mid - 1;
//     else
//       break;
//   }
//   return min;
// }
int findMin(int* nums, int numsSize) {
  int left = 0, mid, right = numsSize - 1;
  while (right > left) {
    mid = left + (right - left) / 2;
    if (nums[mid] > nums[right])
      left = mid + 1;
    else
      right = mid;
  }
  return nums[left];
}

// @lc code=end
// [4,5,6,7,0,1,2]

File: binary-search-medium/153.寻找旋转排序数组中的最小值.py
#
# @lc app=leetcode.cn id=153 lang=python3
#
# [153] 寻找旋转排序数组中的最小值
#
from typing import List


# @lc code=start
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        if left == right or nums[left] < nums[right]:
            return nums[0]
        while left <= right:
            mid = left + ((right - left) >> 1)
            if nums[mid] > nums[mid + 1]:
                return nums[mid + 1]
            if nums[mid] < nums[mid - 1]:
                return nums[mid]
            if nums[mid] > nums[0]:
                left = mid + 1
            else:
                right = mid - 1
        return -1

    # def findMin(self, nums: List[int]) -> int:
    #     left, right = 0, len(nums) - 1
    #     while right > left:
    #         mid = left + (right - left) // 2
    #         if nums[mid] > nums[right]:
    #             left = mid + 1
    #         else:
    #             right = mid
    #     return nums[left]
# @lc code=end
